File: 11_动态规划/lcs最长公共子序列问题/最长公共子序列/PrefixSubstringLCS.py
# Prefix-Substring LCS
# s的前缀与t的子串的最长公共子序列(LCS)
# https://judge.yosupo.jp/problem/prefix_substring_lcs


from typing import Any, Sequence


def max(a: int, b: int) -> int:
    return a if a > b else b


def min(a: int, b: int) -> int:
    return a if a < b else b


class PrefixSubstringLCS:
    __slots__ = "_dp1"

    def __init__(self, seq1: Sequence[Any], seq2: Sequence[Any]) -> None:
        n1, n2 = len(seq1), len(seq2)
        dp1, dp2 = [[0] * (n2 + 1) for _ in range(n1 + 1)], [[0] * (n2 + 1) for _ in range(n1 + 1)]
        dp1[0] = list(range(n2 + 1))
        for i in range(1, n1 + 1):
            cache1, cache2, cache3 = dp1[i], dp2[i], dp1[i - 1]
            for j in range(1, n2 + 1):
                if seq1[i - 1] == seq2[j - 1]:
                    cache1[j] = cache2[j - 1]
                    cache2[j] = cache3[j]
                else:
                    max_, min_ = cache3[j], cache2[j - 1]
                    if max_ < min_:
                        max_, min_ = min_, max_
                    cache1[j] = max_
                    cache2[j] = min_
        self._dp1 = dp1

    def query(self, a: int, b: int, c: int) -> int:
        """查询 seq1[:a) 和 seq2[b:c) 的最长公共子序列长度."""
        cache = self._dp1[a]
        return sum(cache[i] <= b for i in range(b + 1, c + 1))


if __name__ == "__main__":
    q = int(input())
    s = input()
    t = input()
    ords1, ords2 = [ord(ch) for ch in s], [ord(ch) for ch in t]
    lcs = PrefixSubstringLCS(ords1, ords2)
    for _ in range(q):
        a, b, c = map(int, input().split())
        print(lcs.query(a, b, c))

File: 17_模式匹配/PrefixSubstringLCS.go
// Title: Prefix-Substring LCS
// !字符串s的前缀与字符串t的子串的最长公共子序列长度

package main

import (
	"bufio"
	"fmt"
	"os"
)

func main() {
	// https://judge.yosupo.jp/problem/prefix_substring_lcs
	in := bufio.NewReader(os.Stdin)
	out := bufio.NewWriter(os.Stdout)
	defer out.Flush()

	var q int
	fmt.Fscan(in, &q)
	var s, t string
	fmt.Fscan(in, &s, &t)

	LCS := NewPrefixSubstringLCSWithString(s, t)
	for i := 0; i < q; i++ {
		var a, b, c int
		fmt.Fscan(in, &a, &b, &c)
		fmt.Println(LCS.Query(a, b, c))
	}
}

type PrefixSubstringLCS struct {
	dp1 [][]uint32
}

func NewPrefixSubstringLCS(seq1, seq2 []int) *PrefixSubstringLCS {
	n1, n2 := uint32(len(seq1)), uint32(len(seq2))
	dp1, dp2 := make([][]uint32, n1+1), make([][]uint32, n1+1)
	for i := range dp1 {
		dp1[i] = make([]uint32, n2+1)
		dp2[i] = make([]uint32, n2+1)
	}
	for j := uint32(0); j < n2+1; j++ {
		dp1[0][j] = j
	}

	for i := uint32(1); i < n1+1; i++ {
		for j := uint32(1); j < n2+1; j++ {
			if seq1[i-1] == seq2[j-1] {
				dp1[i][j] = dp2[i][j-1]
				dp2[i][j] = dp1[i-1][j]
			} else {
				dp1[i][j] = maxUint32(dp1[i-1][j], dp2[i][j-1])
				dp2[i][j] = minUint32(dp1[i-1][j], dp2[i][j-1])
			}
		}
	}

	return &PrefixSubstringLCS{dp1: dp1}
}

func NewPrefixSubstringLCSWithString(s, t string) *PrefixSubstringLCS {
	ords1, ords2 := make([]int, len(s)), make([]int, len(t))
	for i := range s {
		ords1[i] = int(s[i])
	}
	for i := range t {
		ords2[i] = int(t[i])
	}
	return NewPrefixSubstringLCS(ords1, ords2)
}

// 求 seq1[:a) 和 seq2[b:c) 的最长公共子序列长度.
func (ps *PrefixSubstringLCS) Query(a int, b, c int) int {
	b32, c32 := uint32(b), uint32(c)
	res := 0
	for i := b32 + 1; i < c32+1; i++ {
		if ps.dp1[a][i] <= b32 {
			res++
		}
	}
	return res
}

func minUint32(a, b uint32) uint32 {
	if a < b {
		return a
	}
	return b
}

func maxUint32(a, b uint32) uint32 {
	if a > b {
		return a
	}
	return b
}
